### PR TITLE
Stabilize `lint_reasons` in rustc (RFC 2383)

### PR DESCRIPTION
**Pull request by xFrednet**
The day has come, this PR has the necessary changes to stabilize the `lint_reasons` feature. This covers all changes required in rustc and connected tools. The documentation will be added in a separate PR.

I'll keep this PR as a draft, until the stabilization report has been created in the tracking issue.

---

r? @wesleywiser

cc: https://github.com/rust-lang/rust/issues/54503

### CODE REVIEW

**Comment by xFrednet** [comment]
It seems like Clippy expects the feature to still be present :thinking:. Locally, `./x.py test Clippy` passed just fine... Does somebody know how to fix this error?

**Comment by xFrednet** [comment]
Hey @wesleywiser, @flip1995, the CI is currently failing, since Clippy requires the feature to be complied, even if the feature has been removed. I'm unsure how to fix this. Does somebody know how to fix this or who I can ask about this? :upside_down_face:

**Comment by jyn514** [comment]
@xFrednet those errors are from the stage 0 compiler. You need to add `cfg(bootstrap)` to the feature attribute rather than removing it altogether

**Comment by xFrednet** [comment]
Thank you for the response @jyn514! I've thought about it, but wasn't sure since I expected that clippy is only built with the current rustc version. I've tried adding `#![cfg_attr(feature = "bootstrap", feature(lint_reasons))]` but that fails the local build. `#![cfg_attr(not(feature = "bootstrap"), feature(lint_reasons))]` seams to work, I'll test it here in the CI :upside_down_face:

**Comment by jyn514** [comment]
@xFrednet no, I mean `cfg(bootstrap)`, not `cfg(feature = "bootstrap")`. Your cfg is always false and will fail stage 1 builds.

**Comment by xFrednet** [comment]
Ahh, sorry, then I misunderstood you at first. :sweat_smile:  Now I've changed it to `#![cfg_attr(bootstrap, feature(lint_reasons))]` :upside_down_face:

**Comment by xFrednet** [comment]
The CI is, thank you so much for the support @jyn514! :heart:

File: src/tools/clippy/tests/ui/macro_use_imports.stderr [hunk 1]
```diff
@@ -1,28 +1,28 @@
 error: `macro_use` attributes are no longer needed in the Rust 2018 edition
-  --> $DIR/macro_use_imports.rs:23:5
+  --> $DIR/macro_use_imports.rs:24:5
    |
 LL |     #[macro_use]
-   |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mac::{inner::foofoo, inner::try_err};`
+   |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mac::inner::nested::string_add;`
```
Comment (flip1995) [inline src/tools/clippy/tests/ui/macro_use_imports.stderr:5]: Why did that change? :thinking:

File: src/test/ui/lint/rfc-2383-lint-reason/no_ice_for_partial_compiler_runs.stdout [hunk 2]
```diff
@@ -1,14 +1,12 @@
...
-// This ensures that ICEs like rust#94953 don't happen
-// check-pass
-// compile-flags: -Z unpretty=expanded
-
-#![feature(lint_reasons)]
 #[prelude_import]
 use ::std::prelude::rust_2015::*;
 #[macro_use]
 extern crate std;
+// This ensures that ICEs like rust#94953 don't happen
```
Comment (flip1995) [inline src/test/ui/lint/rfc-2383-lint-reason/no_ice_for_partial_compiler_runs.stdout:7]: Why was this moved?

File: src/doc/rustc/src/lints/levels.md [hunk 3]
```diff
@@ -33,6 +34,41 @@ But this code violates the `missing_docs` lint.
 These lints exist mostly to be manually turned on via configuration, as we'll
 talk about later in this section.
 
+## expect
+
+Sometimes, it can be helpful to suppress lints, but at the same time ensure that
+the code in question still emits them. The 'expect' level does exactly this, with
+it a lint can be expected. If the lint in question is not an emitted, a new lint
+`unfulfilled_lint_expectation` at the attribute notifying you that the expectation,
+is no longer fulfilled.
```
Comment (flip1995) [inline src/doc/rustc/src/lints/levels.md:43]:
```suggestion
the code in question still emits them. The 'expect' level does exactly this. If the lint in question is not  emitted, the
`unfulfilled_lint_expectation` lint triggers on the `expect` attribute, notifying you that the expectation
is no longer fulfilled.
```

this suggestion is not at all formatted.

File: src/doc/rustc/src/lints/levels.md [hunk 4]
```diff
@@ -33,6 +34,41 @@ But this code violates the `missing_docs` lint.
...
+
+```rust
+fn main() {
+    #[expect(unused_variables)]
+    let unused = "Everyone ignores me";
+
+    #[expect(unused_variables)]
+    let used = "I'm usefull";
```
Comment (flip1995) [inline src/doc/rustc/src/lints/levels.md:51]:
```suggestion
    let used = "I'm useful";
```

File: src/doc/rustc/src/lints/levels.md [hunk 5]
```diff
@@ -33,6 +34,41 @@ But this code violates the `missing_docs` lint.
...
+  |
+7 |     #[expect(unused_variables)]
+  |              ^^^^^^^^^^^^^^^^
+  |
+  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
+```
+
+This level can only be defined via the `#[expect]` attribute and not via the
+console. Lints with the special 'force-warn' lint will still be emitted as usual.
+The fulfillment of expectations is still tracked.
```
Comment (flip1995) [inline src/doc/rustc/src/lints/levels.md:70]: Not sure if this is an implementation detail. I don't think the average Rust user has any idea what this sentence means.

Comment (xFrednet) [reply src/doc/rustc/src/lints/levels.md:70]: Might be, should I remove the sentence completely or rephrase it? It could say: "The lint suppression can be overridden by the `--force-warn` argument. The lint will then be emitted as a warning, but the expectation will still be marked as fulfilled if a lint was emitted."

File: src/doc/rustc/src/lints/levels.md (hunk 4 above)
Comment (xFrednet) [reply src/doc/rustc/src/lints/levels.md:51]: Me and my spelling, a never-ending story. Soon I'll have finished season two, you can preorder a copy if you'd like xD

File: src/test/ui/lint/rfc-2383-lint-reason/no_ice_for_partial_compiler_runs.stdout (hunk 2 above)
Comment (xFrednet) [reply src/test/ui/lint/rfc-2383-lint-reason/no_ice_for_partial_compiler_runs.stdout:7]: IDK, I just ran rustc and blessed the tests. I'm guessing that something related to this was merged meanwhile? The previous position put in the prelude, which seems wrong now that I see it :sweat_smile: If the tests pass I think it's fine.

File: src/tools/clippy/tests/ui/macro_use_imports.stderr (hunk 1 above)
Comment (xFrednet) [reply src/tools/clippy/tests/ui/macro_use_imports.stderr:5]: Another one I'm unsure about. The CI also failed, so it might be related to that. I'll try to fix it now.

Comment (flip1995) [reply src/tools/clippy/tests/ui/macro_use_imports.stderr:5]: Hm, I think the order of the lint emission just changed. Which is really weird, but doesn't break tests, so I'm ok with it.

**Review by flip1995 — APPROVED** [review]
Besides the small update in the `levels.md` file, this all LGTM. Not sure if this has to go through an FCP still or what the process here is.

**Comment by xFrednet** [comment]
I'm going to wait with the rebase until the discussion in the tracking issue was resolved :)

**Comment by est31** [comment]
:wave: Hello, I'm writing this comment in this **stabilization PR** to notify you, the authors of this PR, that https://github.com/rust-lang/rust/pull/100591 has been merged, which implemented a change in how features are stabilized.

Your PR has been filed before the change, so will likely require modifications in order to comply with the new rules. I recommend you to:

1. rebase the PR onto latest master, so that uses of the placeholder are possible.
2. replace the version numbers in the PR with the placeholder `CURRENT_RUSTC_VERSION`. For language changes, this means the version numbers in `accepted.rs` (example: 4caedbae869855a19ca49455e2cd338ae3b8c1ef). For library changes, this means the `since` fields (example e576a9b554f24a26c36c04b41c901636d190605b).

That's it! The `CURRENT_RUSTC_VERSION` placeholder will, as part of the release process, be replaced with the version number that the PR merged for. It can be used anywhere in `rust-lang/rust`, not just accepted.rs and the since fields.

If you have any questions, feel free to drop by the [zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/241545-t-release/topic/libs.20stabilization.20placeholder), or ping me directly in this PR's thread. Thanks! :wave:

**Comment by apiraino** [comment]
Switching to waiting on author to incorporate changes (based on [this comment](https://github.com/rust-lang/rust/pull/99063#issuecomment-1229259208)). Feel free to request a review with `@rustbot ready`, thanks!

@rustbot author

**Comment by wesleywiser** [comment]
I believe this is still waiting on consensus in the tracking issue so marking as blocked. 

@xFrednet feel free to ping to me for reviews when consensus emerges! 🙂

**Comment by Nyabinary** [comment]
Wha the update on this?

**Comment by xFrednet** [comment]
The stabilization is awaiting a decision from that lang team, about the mental model of the `#[expect]` attribute. The issue is tagged, we just have to wait until they find the time to give feedback on the issue.

See: https://github.com/rust-lang/rust/issues/54503

**Comment by xFrednet** [comment]
I'm closing this in favor of https://github.com/rust-lang/rust/pull/120924

Thank you for everyone, that participated in this discussion!
